harness/determined/_searcher.py

```diff
@@ -0,0 +1,145 @@
+import enum
+import logging
+from typing import Iterator, Optional
+
+import determined as det
+from determined.experimental import client
+
+log = logging.getLogger("generic")
+
+
+class Unit(enum.Enum):
+    EPOCHS = "UNIT_EPOCHS"
+    RECORDS = "UNIT_RECORDS"
+    BATCHES = "UNIT_BATCHES"
+
+
+class SearcherOp:
+    def __init__(
+        self,
+        session: client.Session,
+        trial_id: int,
+        unit: Unit,
+        length: int,
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._unit = unit
+        self._length = length
+        self._completed = False
+
+    @property
+    def unit(self) -> Unit:
+        return self._unit
+
+    @property
+    def length(self) -> int:
+        return self._length
+
+    @property
+    def records(self) -> int:
+        if self._unit != Unit.RECORDS:
+            raise AssertionError(
+                "you can only read op.records if you configured your searcher in terms of records"
+            )
+        return self._length
+
+    @property
+    def batches(self) -> int:
+        if self._unit != Unit.BATCHES:
+            raise AssertionError(
+                "you can only read op.batches if you configured your searcher in terms of batches"
+            )
+        return self._length
+
+    @property
+    def epochs(self) -> int:
+        if self._unit != Unit.EPOCHS:
+            raise AssertionError(
+                "you can only read op.epochs if you configured your searcher in terms of epochs"
+            )
+        return self._length
+
+    def report_progress(self, length: float) -> None:
```

Comment:
this is reporting progress on a single searcher op and not progress of the whole hp search experiment?  what is single op progress used for?

Reply:
With push arch, all ops are absolute lengths.  So if your first op is `10 epochs`, you could report progress as `5` to indicate you're halfway done with that op.  You don't know if this op is your last op (the master might not even know).

But imagine that you finish 10 epochs, complete the op, and then get a second op for `20 epochs`.  Since ops are absolute lengths, after one more epoch of training you would report progress as `11`.  If the master happened to know that this was the only trial and this was the last op, then total progress in the webui would be `55%`.